cli/src/validate.rs

```diff
@@ -0,0 +1,261 @@
+    }
+    let report_validations: Vec<(JunitReportValidation, String)> = reports
+        .into_iter()
+        .map(|report| (validate_report(&report.0), report.1))
+        .collect();
+
+    // print results
+    let (num_invalid_reports, num_suboptimal_reports) =
+        print_validation_errors(&report_validations);
+    if num_invalid_reports == 0 {
+        print_summary_success(report_validations.len(), num_suboptimal_reports);
+        Ok(EXIT_SUCCESS)
+    } else {
+        print_summary_failure(
+            report_validations.len(),
+            num_invalid_reports,
+            num_suboptimal_reports,
+        );
+        Ok(EXIT_FAILURE)
+    }
+}
+
+fn parse_file_sets(
+    file_sets: Vec<FileSet>,
+) -> anyhow::Result<(Vec<(Report, String)>, Vec<(JunitParseError, String)>)> {
+    file_sets.iter().try_fold(
+        (
+            Vec::<(Report, String)>::new(),          // Vec<(Report, file path)>
+            Vec::<(JunitParseError, String)>::new(), // Vec<(JunitParseError, file path)>
+        ),
+        |mut file_sets_parse_results, file_set| {
+            let file_set_parse_results = file_set.files.iter().try_fold(
+                (
+                    Vec::<(Report, String)>::new(),
+                    Vec::<(JunitParseError, String)>::new(),
+                ),
+                |mut file_set_parse_results, bundled_file| {
+                    let path = std::path::Path::new(&bundled_file.original_path);
+                    let file = std::fs::File::open(path)?;
+                    let file_buf_reader = BufReader::new(file);
+                    let mut junit_parser = JunitParser::new();
+                    junit_parser.parse(file_buf_reader).context(format!(
+                        "Encountered unrecoverable error while parsing file: {}",
+                        bundled_file.original_path_rel
+                    ))?;
+                    file_set_parse_results.1.extend(
+                        junit_parser
+                            .errors()
+                            .iter()
+                            .map(|e| (*e, bundled_file.original_path_rel.clone())),
+                    );
+                    file_set_parse_results.0.extend(
+                        junit_parser
+                            .into_reports()
+                            .iter()
+                            .map(|report| (report.clone(), bundled_file.original_path_rel.clone())),
+                    );
+                    Ok::<(Vec<(Report, String)>, Vec<(JunitParseError, String)>), anyhow::Error>(
+                        file_set_parse_results,
+                    )
+                },
+            )?;
+            file_sets_parse_results.0.extend(file_set_parse_results.0);
+            file_sets_parse_results.1.extend(file_set_parse_results.1);
+            Ok::<(Vec<(Report, String)>, Vec<(JunitParseError, String)>), anyhow::Error>(
+                file_sets_parse_results,
+            )
```

Comment:
If you set a return type of the closure you won't have to annotate the type of `Result` here:

e.g.

```rs
        |mut file_sets_parse_results, file_set| -> anyhow::Result<(Vec<(Report, String)>, Vec<(JunitParseError, String)>)> {
            // ...
            Ok(file_sets_parse_results)
        }
```